File: week3/q2.c
#include <stdio.h>
#include <stdlib.h>

struct Node { int data; struct Node* next; };

struct Node* createNode(int val) {
    struct Node* temp = malloc(sizeof(struct Node));
    temp->data = val;
    temp->next = NULL;
    return temp;
}

struct Node* insertEnd(struct Node* head, int val) {
    struct Node* newNode = createNode(val);
    if (!head) return newNode;
    struct Node* curr = head;
    while (curr->next) curr = curr->next;
    curr->next = newNode;
    return head;
}

struct Node* insertMiddle(struct Node* head, int pos, int val) {
    struct Node* newNode = createNode(val);
    struct Node* curr = head;
    for (int i=1; i<pos-1 && curr; i++) curr = curr->next;
    newNode->next = curr->next;
    curr->next = newNode;
    return head;
}

void printList(struct Node* head) {
    while (head) { printf("%d ", head->data); head = head->next; }
}

int main() {
    int n, pos, val, x;
    scanf("%d", &n);
    struct Node* head = NULL;
    for (int i=0; i<n; i++) { scanf("%d", &x); head = insertEnd(head, x); }
    scanf("%d%d", &pos, &val);
    head = insertMiddle(head, pos, val);
    printList(head);
}

File: week3/q2.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class LinkedList:
    def __init__(self):
        self.head = None

    def insert_end(self, val):
        new_node = Node(val)
        if not self.head:
            self.head = new_node
            return
        curr = self.head
        while curr.next:
            curr = curr.next
        curr.next = new_node

    def insert_middle(self, pos, val):
        new_node = Node(val)
        curr = self.head
        for _ in range(pos-2):
            curr = curr.next
        new_node.next = curr.next
        curr.next = new_node

    def display(self):
        curr = self.head
        while curr:
            print(curr.data, end=" ")
            curr = curr.next

n = int(input())
vals = list(map(int, input().split()))
pos, val = map(int, input().split())
ll = LinkedList()
for v in vals:
    ll.insert_end(v)
ll.insert_middle(pos, val)
ll.display()
